doc/designs/quic-design/synthesis.md

```diff
@@ -0,0 +1,494 @@
+ * These are independent of one another.
+ *
+ * If write-multiple-datagram and read-multiple-datagram BIO APIs are used. the
+ * source/destination information is passed in structures used by those APIs.
+ * The state configured by these functions is then undefined.
+ *
+ * For a BIO_dgram_mem, these functions will fail if the necessary capability
+ * flags have not been set using BIO_set_dgram_caps. These functions are always
+ * available for a BIO_dgram.
+ *
+ * These functions are inspired by #5257 but differ, because they use separate
+ * state for the RX and TX side. This enables RX and TX to be performed from
+ * different threads if desired. They have been renamed to reflect this and be
+ * less confusing. Alternately, the read/write-multiple APIs can be used which
+ * avoid mutating state inside the BIO_dgram, allowing arbitrary multithreaded
+ * use.
+ */
+__owur int BIO_get_dgram_rx_src_addr(BIO *bio, BIO_ADDR *addr);
+    /* BIO_CTRL_GET_DGRAM_RX_SRC_ADDR */
+__owur int BIO_get_dgram_rx_dst_addr(BIO *bio, BIO_ADDR *addr);
+    /* BIO_CTRL_GET_DGRAM_RX_DST_ADDR */
+
+__owur int BIO_set_dgram_tx_src_addr(BIO *bio, const BIO_ADDR *addr);
+    /* (non-MVP) BIO_CTRL_SET_DGRAM_TX_SRC_ADDR */
+__owur int BIO_set_dgram_tx_dst_addr(BIO *bio, const BIO_ADDR *addr);
+    /* BIO_CTRL_SET_DGRAM_TX_DST_ADDR */
+
+/*
+ * Read/write multiple API.
+ *
+ * This is evolved from my initial proposal in #18210 as well as from Matt's
+ * original QUIC proposal.
+ *
+ * This features:
+ *   - An intentionally high level message structure which only supports
+ *     the functionality we want to support and which we can support
+ *     across a wide range of platforms.
+ *
+ *   - No iovec support, as this cannot be performantly supported on some
+ *     platforms.
+ *
+ *   - Ability to get both source and destination addresses both for RX and TX.
+ *
+ *   - Limited extensibity via per-message and global flags arguments.
+ *
+ * The caller sets data to point to the data buffer, and peer and local to
+ * point to BIO_ADDRs to be filled with the peer and local addresses.
+ * If peer or local are NULL, this data is simply not returned.
+ *
+ * data_len is set to the data buffer length on call and has the
+ * actual amount of data sent/received written to it afterwards,
+ * assuming the message was touched at all, as determined by the return
+ * value of the function:
+ *
+ * These functions return the number n of messages sent/received, which will
+ * always be less than or equal to num_msg, or -1:
+ *
+ *   Postcondition: n <= num_msg
+ *
+ *   Postcondition: msg[i] for all 0 <= i < n have had their data_len
+ *      modified to the actual amount of data sent or received.
+ *
+ * The flags field of each message defines per-message flags and is used for
+ * both input and output. If a message is processed, this field is written
+ * with any result flags ("message truncated", etc.)
+ *
+ * There are no per-message flags currently defined.
```

Comment:
We might need for example the ECN flag (not for MVP of course).

Reply:
Yeah, definitely.